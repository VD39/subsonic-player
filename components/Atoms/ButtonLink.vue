<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
