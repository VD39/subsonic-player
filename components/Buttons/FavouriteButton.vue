<script setup lang="ts">
import IconButton from './IconButton.vue';

const props = defineProps<{
  favourite: boolean;
  id: string;
  type: MediaType;
}>();

const { addFavourite, removeFavourite } = useFavourite();

const isFavourite = ref(props.favourite);

const buttonProps = computed<FavouriteButtonProps>(() => {
  const message = isFavourite.value ? 'Unlike' : 'Like';

  return {
    iconColor: isFavourite.value ? 'var(--error-color)' : undefined,
    iconWeight: isFavourite.value ? 'fill' : 'duotone',
    text: `${message} ${props.type}`,
  };
});

async function toggleFavourite() {
  const action = isFavourite.value ? removeFavourite : addFavourite;

  await action(props);

  isFavourite.value = !isFavourite.value;
}
</script>

<template>
  <IconButton
    icon="PhHeart"
    :icon-color="buttonProps.iconColor"
    :icon-weight="buttonProps.iconWeight"
    :title="buttonProps.text"
    v-bind="$attrs"
    @click="toggleFavourite"
  >
    {{ buttonProps.text }}
  </IconButton>
</template>
