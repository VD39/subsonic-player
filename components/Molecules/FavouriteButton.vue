<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

const props = defineProps<{
  favourite: boolean;
  id: string;
  type: MediaType;
}>();

const { addToFavouriteIds, favouriteIds, toggleFavourite } = useFavourite();

const isFavourite = computed(() => !!favouriteIds.value[props.id]);

const buttonProps = computed<ButtonProps>(() => ({
  iconColor: isFavourite.value ? 'var(--error-color)' : undefined,
  iconWeight: isFavourite.value ? 'fill' : 'regular',
  text: `${isFavourite.value ? 'Unlike' : 'Like'} ${props.type}`,
}));

function toggleCurrentFavourite() {
  // No need to pass isFavourite.value value, use props directly.
  toggleFavourite(props, isFavourite.value);
}

watch(
  () => props.id || props.favourite,
  () => {
    // Check if id exists as a key, cannot check value as it could be false.
    if (props.id in favouriteIds.value) {
      return;
    }

    if (props.favourite) {
      addToFavouriteIds(props.id);
    }
  },
  {
    immediate: true,
  },
);
</script>

<template>
  <ButtonLink
    :icon="ICONS.favourite"
    :icon-color="buttonProps.iconColor"
    :icon-weight="buttonProps.iconWeight"
    :title="buttonProps.text"
    @click="toggleCurrentFavourite"
  >
    {{ buttonProps.text }}
  </ButtonLink>
</template>
