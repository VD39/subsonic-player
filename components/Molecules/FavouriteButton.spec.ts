import type { VueWrapper } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import FavouriteButton from './FavouriteButton.vue';

const addFavouriteMock = vi.fn();
const removeFavouriteMock = vi.fn();

mockNuxtImport('useFavourite', () => () => ({
  addFavourite: addFavouriteMock,
  removeFavourite: removeFavouriteMock,
}));

function factory(props = {}) {
  return mount(FavouriteButton, {
    props: {
      favourite: false,
      id: 'id',
      type: 'album',
      ...props,
    },
    slots: {
      default: 'Default slot content.',
    },
  });
}

const buttonProps = {
  false: {
    color: undefined,
    mockFn: addFavouriteMock,
    title: 'Like',
    weight: 'duotone',
  },
  true: {
    color: 'var(--error-color)',
    mockFn: removeFavouriteMock,
    title: 'Unlike',
    weight: 'fill',
  },
};

describe('FavouriteButton', () => {
  let wrapper: VueWrapper;

  describe.each([
    [false, buttonProps.false, buttonProps.true],
    [true, buttonProps.true, buttonProps.false],
  ])(
    'when favourite prop is set to %s',
    (favourite, currentProps, inverseProp) => {
      beforeEach(() => {
        wrapper = factory({
          favourite,
        });
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('sets the correct color on ButtonLink component', () => {
        expect(wrapper.findComponent(ButtonLink).props('iconColor')).toBe(
          currentProps.color,
        );
      });

      it('sets the correct weight on ButtonLink component', () => {
        expect(wrapper.findComponent(ButtonLink).props('iconWeight')).toBe(
          currentProps.weight,
        );
      });

      it('sets the correct title on ButtonLink component', () => {
        const title = `${currentProps.title} album`;
        const buttonLink = wrapper.findComponent(ButtonLink);

        expect(buttonLink.attributes('title')).toBe(title);
        expect(buttonLink.text()).toBe(title);
      });

      describe('when the ButtonLink component emits a click event', () => {
        beforeEach(() => {
          wrapper.findComponent(ButtonLink).vm.$emit('click');
        });

        it('matches the snapshot', () => {
          expect(wrapper.html()).toMatchSnapshot();
        });

        it('calls the correct favourite action', () => {
          expect(currentProps.mockFn).toHaveBeenCalled();
        });

        it('sets the correct color on ButtonLink component', () => {
          expect(wrapper.findComponent(ButtonLink).props('iconColor')).toBe(
            inverseProp.color,
          );
        });

        it('sets the correct weight on ButtonLink component', () => {
          expect(wrapper.findComponent(ButtonLink).props('iconWeight')).toBe(
            inverseProp.weight,
          );
        });

        it('sets the correct title on ButtonLink component', () => {
          const title = `${inverseProp.title} album`;
          const buttonLink = wrapper.findComponent(ButtonLink);

          expect(buttonLink.attributes('title')).toBe(title);
          expect(buttonLink.text()).toBe(title);
        });
      });
    },
  );
});
