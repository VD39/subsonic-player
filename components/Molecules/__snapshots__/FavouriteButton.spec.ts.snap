// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavouriteButton > when favouriteIds id key is false > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" type="button" title="Like album">
  <phheart aria-hidden="true" class="icon" size="24" weight="regular"></phheart><span class="clamp text visuallyHidden">Like album</span>
</button>"
`;

exports[`FavouriteButton > when favouriteIds id key is true > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" type="button" title="Unlike album">
  <phheart aria-hidden="true" class="icon" color="var(--error-color)" size="24" weight="fill"></phheart><span class="clamp text visuallyHidden">Unlike album</span>
</button>"
`;
